rviz_simulator/include/rviz_simulator/camera_calibration_optimizer.h

```diff
@@ -0,0 +1,207 @@
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the Willow Garage, Inc. nor the names of its
+ *       contributors may be used to endorse or promote products derived from
+ *       this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Christopher Sahadeo
+ *
+ * Ceres optimizer for a bundle adjustment camera calibration problem
+ * Considers the reprojection error to optimize the following parameters:
+ *    world_T_camera
+ *    world_T_target
+ *    camera_intrinsics
+ *
+ */
+
+#ifndef RVIZ_SIMULATOR_CAMERA_CALIBRATION_OPTIMIZER_H
+#define RVIZ_SIMULATOR_CAMERA_CALIBRATION_OPTIMIZER_H
+
+#include <ros/ros.h>
+#include <ros/package.h>
+
+#include <fstream>
+
+#include "ceres/ceres.h"
+#include "ceres/rotation.h"
+
+#include <yaml-cpp/yaml.h>
+
+#include <Eigen/Dense>
+
+// for getting a directory/file listing
+#include <dirent.h>
+
+// for making directories
+#include <sys/stat.h>
+#include <sys/types.h>
+
+// for outputting camera calibration properties to YAML
+#include <camera_calibration_parsers/parse_yml.h>
+
+namespace camera_calibration
+{
+#define REFERENCE_T_TARGET_SIZE 6
+#define OBJ_POINTS_SIZE 3
+#define CORNER_POINTS_SIZE 2
+#define NUM_OBJ_POINTS 4
+#define NUM_CORNERS 4
+#define CAMERA_INTRINSICS_SIZE 9  // currently hardcoded for the plumb bob model
+
+// All transforms are stored as a 6 elemnt double array
+// [0, 1, 2] - rodriguez angle-axis rotation
+// [3, 4, 5] - translation
+
+/// The target struct models a fiducial target (E.g. an AprilTag)
+/// It stores the world_T_target transform
+/// and the 3D object point coordinates in the target frame
+struct Target
+{
+  int targetID;
+  std::array<double, REFERENCE_T_TARGET_SIZE> world_T_target;
+  std::array<std::array<double, OBJ_POINTS_SIZE>, NUM_OBJ_POINTS> obj_points_in_target;
```

Comment:
I used `array` over `vector` since the number of object points in the target frame is known and fixed (4)

I did not use `geometry_msgs::Point` since it requires additional, _confusing_, **_painful_** parametrization to get the ceres cost function working; a project worthy of a separate PR in the future.